src/view/renderer.js

```diff
@@ -639,6 +652,42 @@ export default class Renderer {
 			}
 		}
 	}
+
+	/**
+	 * Checks if two given selections are similar. Selections are considered similar if they are non-collapsed
+	 * and their trimmed (see {@link #_trimSelection}) representations are equal.
+	 *
+	 * @private
+	 * @param {module:engine/view/selection~Selection} current
+	 * @param {module:engine/view/selection~Selection} previous
+	 * @returns {Boolean}
+	 */
+	_areSimilarSelections( current, previous ) {
+		return !current.isCollapsed && this._trimSelection( current ).isEqual( this._trimSelection( previous ) );
+	}
+
+	/**
+	 * Creates a copy of a given selection with all of its ranges
+	 * trimmed (see {@link module:engine/view/range~Range#getTrimmed getTrimmed}).
+	 *
+	 * @private
+	 * @param {module:engine/view/selection~Selection} selection
+	 * @returns {module:engine/view/selection~Selection} Selection copy with all ranges trimmed.
+	 */
+	_trimSelection( selection ) {
+		const newSelection = Selection.createFromSelection( selection );
```

Comment:
Let me add that this is all due to our lack of trust that no one will touch the ranges and positions.

In a perfect immutable world, we'd only need to copy new instances on `getTrimmed()`. No other method would need to do that because no other method change ranges. And I think that we must go into this direction. Otherwise, we create a huge memory pressure.